### PR TITLE
Added unsupervised tokenization model

### PR DESCRIPTION
**Pull request by zredlined**
#2 adding support for sentencepiece tokenizer, PR for initial code review.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,3 +1,4 @@
 tensorflow==2.1.0
 tensorflow_privacy==0.2.2
+sentencepiece
```
Comment (johntmyers) [inline requirements.txt]: Maybe we should pin the version to future proof a bit

File: setup.py [hunk 2]
```diff
@@ -21,6 +21,7 @@
     install_requires=[
         'tensorflow==2.1.0',
         'tensorflow_privacy==0.2.2',
-        'smart_open'
+        'smart_open',
+        'sentencepiece'
```
Comment (johntmyers) [inline setup.py]: Same here re: pinning the version

File: examples/notebook_utils.py [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+
+def set_working_dir() -> str:
+    # Change to project src directory
+    working_dir = Path.cwd()
+    if working_dir.parts[-1] == 'examples':
+        working_dir = Path(Path.cwd()).parents[0] / 'src'
+        os.chdir(working_dir)
+
+    print(working_dir.resolve())
```
Comment (johntmyers) [inline examples/notebook_utils.py]: Can this be removed?

File: tests/conftest.py [hunk 4]
```diff
@@ -1,10 +1,21 @@
 from unittest.mock import patch
 import pytest
+import os
+from pathlib import Path
+import shutil
 
 from gretel_synthetics.config import LocalConfig
+from gretel_synthetics.train import annotate_training_data
 
+test_data_dir = os.path.join(str(Path(__file__).parent.resolve()), 'data')
```
Comment (misberner) [inline tests/conftest.py]: `str((Path(__file__).parent / 'data').resolve())`, or skip the `str` and use `pathlib` downstream

File: tests/test_config.py [hunk 5]
```diff
@@ -8,8 +8,9 @@
...
-    lc = LocalConfig(checkpoint_dir=target, training_data='blah')
+    lc = LocalConfig(checkpoint_dir=target, input_data='blah')
     mkdir.assert_called_with(target)
     assert lc.epochs == 30
-    assert lc.training_data == 'blah'
-    assert lc.char2idx == os.path.join(target, 'char2idx.p')
\ No newline at end of file
+    assert lc.input_data == 'blah'
+    assert lc.training_data == os.path.join(target, 'training_data.txt')
+    assert lc.tokenizer_model == os.path.join(target, 'm.model')
```
Comment (misberner) [inline tests/test_config.py]: nit: add newline at end of file

File: tests/test_train.py [hunk 6]
```diff
@@ -53,4 +41,4 @@ def test_train_rnn(_open, trng, pickle, model, smol_data, global_local_config):
     )
 
     mock_model.fit.assert_called
-    
\ No newline at end of file
+    """
```
Comment (misberner) [inline tests/test_train.py]: nit: newline at end of file

File: examples/notebook_utils.py (hunk 3 above)
Comment (zredlined) [reply examples/notebook_utils.py]: Yes. Let's go with existing functionality to run against installed library vs. local. removing notebook_utils.py for now

File: src/gretel_synthetics/config.py [hunk 7]
```diff
@@ -47,16 +51,17 @@ def _gen_idxs(self):  # pragma: no cover
...
 
-        self._set_idxs()
+        self._set_tokenizer()
 
-    def _set_idxs(self):
-        self.char2idx = os.path.join(self.checkpoint_dir, 'char2idx.p')
-        self.idx2char = os.path.join(self.checkpoint_dir, 'idx2char.p')
+    def _set_tokenizer(self):
+        self.tokenizer_prefix = "m"
+        self.tokenizer_model = os.path.join(self.checkpoint_dir, 'm.model')
```
[the remark this answers is not in the document]
Comment (zredlined) [reply src/gretel_synthetics/config.py]: good call. updated code and mocks to use Pathlib

File: src/gretel_synthetics/train.py [hunk 8]
```diff
@@ -70,45 +60,64 @@ def train_rnn(store: BaseConfig):
         logging.info('Trained with non-private Adam optimizer')
 
 
-def create_dataset(store: BaseConfig, text: str, vocab: list) -> tf.data.Dataset:
+def annotate_training_data(store: BaseConfig):
+    # required for sentencepiece to tokenize newline characters
+    logging.info(f"Annotating training data from {store.input_data}")
+    labeled_text = open(store.input_data, 'r').read().replace('\n', '<n>\n')
```
[the remark this answers is not in the document]
Comment (zredlined) [reply src/gretel_synthetics/train.py]: went w/cleaner implementation reading line by line

**Pull request merged**